### PR TITLE
Add arbitrary replaceWhere functionality

### PR DESCRIPTION
**Pull request by SpaceRangerWes**
**Feature Request:**
Create ability to selectively overwrite only the data that matches predicates over arbitrary columns by enhancing the replaceWhere function.

**Requirements:**
Given a set of column and values for them from an input DataFrame, remove all rows in a target Delta table that meet that condition and replace them with the rows from the input DataFrame. Note that the record matches do not have to be unique.

**Why is this different than upsert?** With upsert existing rows are changed and new rows are inserted. This function not only requires changes and inserts, but also deletes.

**Why can’t the existing replaceWhere function be used?** replaceWhere is only good in the case of replacing complete partitions. Not individual records belonging to it.

### CODE REVIEW

**Comment by SpaceRangerWes** [comment]
@marmbrus / @brkyvz I've opened a new PR for this feature. I had a bunch of conflict on my local system so just started fresh. Apologies for the inconvenience.

**Comment by brkyvz** [comment]
Hey @SpaceRangerWes. Can the `replaceWhere` operation be phrased as a:
```sql
MERGE INTO t
USING s
ON replaceWhereConditionAllOnTarget
WHEN MATCHED THEN DELETE
WHEN NOT MATCHED THEN INSERT *
```
If that works, I think we can avoid a lot of work, and actually add a Catalyst rule which calls `WriteIntoDelta` when all `replaceWhere` references refer to partition columns, and pass to `MERGE` when data columns are involved. cc @tdas ?

**Comment by brkyvz** [comment]
Actually, thinking about it more, we also want replaceWhere to perform checks for us, that the data being inserted also has the exact same values in the replaceWhere condition, so we can't just delegate to MERGE.

**Comment by SpaceRangerWes** [comment]
@brkyvz In a prior PR, @marmbrus said w.r.t. the API

> I haven't had time to review the code closely, but I have one high level comment on the API. I would probably avoid having two different APIs, but would instead extend the existing replaceWhere API to be more powerful. This is more inline with how other features in Delta are implemented (i.e. DELETE).
> 
> We can split the predicates to see if they reference only partition columns or if they reference data columns as well and then pick the most efficient execution plan from there.

https://github.com/delta-io/delta/pull/148#issuecomment-529650839

In my original design doc, I outlined both approaches so I'm honestly indifferent to whether it is under the existing `replaceWhere` or not.

File: src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala [hunk 1]
```diff
@@ -198,40 +198,81 @@ class DeltaSuite extends QueryTest
...
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, 10), (2, 20), (3, 10), (4, 20), (4, 10)).toDF("key", "value")
+        .write
+        .format("delta")
+        .partitionBy("value")
+        .save(tempDir.getCanonicalPath)
+
+      Seq((4, 30), (5, 20)).toDF("key", "value")
```
Comment (brkyvz) [inline src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala]: Yeah, in order to follow the semantics of replaceWhere, this should throw an exception saying that the new data did not contain the replaceWhere condition.

**Comment by SpaceRangerWes** [comment]
> Hey @SpaceRangerWes. I think the logic you have here makes total sense, but there are a couple potential improvements I'd like to suggest:
> 
> 1. Maybe we can refactor the code in DeleteCommand where it can return us a list of actions instead of just committing a transaction. This way, we wouldn't need any new or duplicated code in WriteIntoDelta.
> 2. We also need to perform a check that we're not writing new data with values other than the `replaceWhere` condition. We should throw an error otherwise. Therefore I'm thinking that we should write the data first, and then delete.
> 
> Then the question is, how do we figure out whether the data we have written out matches our `replaceWhere` criteria? Fortunately, we have the invariant checker that can check arbitrary SQL expressions: https://github.com/delta-io/delta/blob/master/src/main/scala/org/apache/spark/sql/delta/schema/DeltaInvariantCheckerExec.scala
> that will ensure that we're writing data that matches our replaceWhere condition. We can throw a nice error otherwise.
> 
> What do you think?

@brkyvz I think this makes a fair amount of sense. The additional transaction kinda bugged me so I like your direction.

**Pull request closed without merging**